test/test.lua

```diff
@@ -2587,6 +2587,69 @@ function torchtest.storageview()
    mytester:assert(13 == s1[2], "should have 13 at position 1")
 end
 
+function torchtest.nonzero()
+  local nSrc = 12
+
+  local types = {
+      'torch.ByteTensor',
+      'torch.CharTensor',
+      'torch.ShortTensor',
+      'torch.IntTensor',
+      'torch.FloatTensor',
+      'torch.DoubleTensor',
+      'torch.LongTensor',
+  }
+
+  local shapes = {
+      torch.LongStorage{12},
+      torch.LongStorage{12, 1},
+      torch.LongStorage{1, 12},
+      torch.LongStorage{6, 2},
+      torch.LongStorage{3, 2, 2},
+  }
+
+  for _, type in ipairs(types) do
+    local tensor = torch.rand(nSrc):mul(2):floor():type(type)
+      for _, shape in ipairs(shapes) do
+        tensor = tensor:reshape(shape)
+        local dst1 = torch.nonzero(tensor)
+        local dst2 = tensor:nonzero()
+        -- Does not work. Torch uses the first argument to determine what
+        -- type the Tensor is expected to be. In our case the second argument
+        -- determines the type of Tensor.
+        --local dst3 = torch.LongTensor()
+        --torch.nonzero(dst3, tensor)
```

Comment:
It's annoying that this doesn't work, but you should be able to do something like this: `torch.FloatTensor.nonzero(...)`, or for arbitrary types: `torch[type].nonzero(...)`. Or even `tensor.add(dst3, tensor)`?

I think we should document that because it's important to be able to use an existing tensor for the result and it's not obvious how to do that.